server/datastore/mysql/software_installers.go

```diff
@@ -0,0 +1,139 @@
+package mysql
+
+import (
+	"context"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jmoiron/sqlx"
+)
+
+func (ds *Datastore) MatchOrCreateSoftwareInstaller(ctx context.Context, payload *fleet.UploadSoftwareInstallerPayload) (uint, error) {
+	// TODO: confirm ds context is appropriate in each of the following (should we use writer for all?)
+	titleID, err := ds.getOrGenerateSoftwareInstallerTitleID(ctx, payload.Title, payload.Source)
+	if err != nil {
+		return 0, err
+	}
+
+	installScriptID, err := ds.getOrGenerateScriptContentsID(ctx, payload.InstallScript)
+	if err != nil {
+		return 0, err
+	}
+
+	var postInstallScriptID *uint
+	if payload.PostInstallScript != "" {
+		sid, err := ds.getOrGenerateScriptContentsID(ctx, payload.PostInstallScript)
+		if err != nil {
+			return 0, err
+		}
+		postInstallScriptID = &sid
+	}
+
+	var tid uint
+	if payload.TeamID != nil {
+		tid = *payload.TeamID
+	}
+
+	stmt := `
+INSERT INTO software_installers (
+	team_id,
+	global_or_team_id, 
+	title_id, 
+	storage_id,
+	filename, 
+	version,
+	install_script_content_id, 
+	pre_install_query, 
+	post_install_script_content_id
+) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`
+
+	args := []interface{}{
+		payload.TeamID,
+		tid,
+		titleID,
+		payload.StorageID,
+		payload.Filename,
+		payload.Version,
+		installScriptID,
+		payload.PreInstallQuery,
+		postInstallScriptID,
+	}
+
+	res, err := ds.writer(ctx).ExecContext(ctx, stmt, args...)
+	if err != nil {
+		if isDuplicate(err) {
+			// already exists for this team/no team
+			err = alreadyExists("SoftwareInstaller", payload.Title)
+		}
+		return 0, ctxerr.Wrap(ctx, err, "insert software installer")
+	}
+
+	id, _ := res.LastInsertId()
+
+	// TODO: should we try to wrap all of the above in a transaction?
```

Comment:
I'd generally say yes, but it would be hard (with the get or generate) and we don't need to AFAICT - if the first parts work and it fails later on, it will just have created a software title (that will eventually get deleted) or script contents that will also eventually be garbage collected. So I'd leave it like that.